In [ ]:
import pandas as pd

## Reading files

In [2]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [3]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [4]:
wh = pd.concat([wh2018,wh2019])

In [5]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [6]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [7]:
wh

Overall rank         Country or region  Score  GDP per capita  \
0               1                   Finland  7.632           1.305   
1               2                    Norway  7.594           1.456   
2               3                   Denmark  7.555           1.351   
3               4                   Iceland  7.495           1.343   
4               5               Switzerland  7.487           1.420   
..            ...                       ...    ...             ...   
151           152                    Rwanda  3.334           0.359   
152           153                  Tanzania  3.231           0.476   
153           154               Afghanistan  3.203           0.350   
154           155  Central African Republic  3.083           0.026   
155           156               South Sudan  2.853           0.306   

     Social support  Healthy life expectancy  Freedom to make life choices  \
0             1.592                    0.874                         0.681   
1             1.582                    0.861                         0.686   
2             1.590                    0.868                         0.683   
3               NaN                    0.914                         0.677   
4             1.549                    0.927                         0.660   
..              ...                      ...                           ...   
151           0.711                    0.614                         0.555   
152           0.885                    0.499                         0.417   
153           0.517                    0.361                         0.000   
154           0.000                    0.105                         0.225   
155           0.575                    0.295                         0.010   

     Generosity  Perceptions of corruption  year  
0         0.202                      0.393  2018  
1         0.286                      0.340  2018  
2         0.284                      0.408  2018  
3           NaN                      0.138  2018  
4         0.256                      0.357  2018  
..          ...                        ...   ...  
151       0.217                      0.411  2019  
152       0.276                      0.147  2019  
153       0.158                      0.025  2019  
154       0.235                      0.035  2019  
155       0.202                      0.091  2019  

[312 rows x 10 columns]

In [8]:
DFNULL= wh.isnull().sum()/wh.shape[0];
DFPER = DFNULL*100

DFPER

Overall rank                    0.000000
Country or region               0.000000
Score                           6.089744
GDP per capita                  3.846154
Social support                  5.128205
Healthy life expectancy         8.333333
Freedom to make life choices    7.371795
Generosity                      6.730769
Perceptions of corruption       4.166667
year                            0.000000
dtype: float64

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [9]:
DFY = wh.groupby('year');
RESUMEN = DFY.describe();
RESUMEN

Overall rank                                                    Score  \
            count  mean        std  min    25%   50%     75%    max  count   
year                                                                         
2018        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  151.0   
2019        156.0  78.5  45.177428  1.0  39.75  78.5  117.25  156.0  142.0   

                ... Generosity        Perceptions of corruption            \
          mean  ...        75%    max                     count      mean   
year            ...                                                         
2018  5.354060  ...    0.23400  0.598                     149.0  0.112859   
2019  5.434028  ...    0.24825  0.566                     150.0  0.109053   

                                                     
           std  min      25%    50%      75%    max  
year                                                 
2018  0.098108  0.0  0.05000  0.082  0.13800  0.457  
2019  0.093279  0.0  0.04775  0.085  0.14025  0.453  

[2 rows x 64 columns]

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

## resolucion ejercicio tres

In [13]:
for i in wh.columns[wh.isnull().any(axis=0)]:  
    wh[i].fillna(wh[i].mean(),inplace=True)
    
DFPR = wh.loc[wh["year"] == wh.fillna] = DFY.mean()
DFPR

Overall rank     Score  GDP per capita  Social support  \
year                                                           
2018          78.5  5.355302        0.892984        1.206033   
2019          78.5  5.430330        0.900683        1.208257   

      Healthy life expectancy  Freedom to make life choices  Generosity  \
year                                                                      
2018                 0.602365                      0.454157    0.181572   
2019                 0.717299                      0.397587    0.185046   

      Perceptions of corruption  
year                             
2018                   0.112773  
2019                   0.109126

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [11]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro